### PR TITLE
#1 Drag and drop support

### PR DESCRIPTION
**Pull request by xiio**
Support for drag and drop tabs between groups

### CODE REVIEW

**Comment by fllp** [comment]
If it helps speeding up the merge: I just tested this. Works fine! (on FF 45 x64 on Linux)

**Comment by skyschub** [comment]
The code still fails eslint. Please rebase your code to the latest `develop`-branch (`git rebase develop`) and run `jake lint` and fix the violations... I'll probably set up an automatic style guide bot for future PRs, but you should do the checks yourself anyway.

**Comment by slowly-sloth** [comment]
I hope it should work now... Is this ok, that for your unchanged code "jake lint" returns me
`? 42 problems (15 errors, 27 warnings)` ?

**Comment by skyschub** [comment]
The `develop` code passes with warnings for the unused modules and two long lines. I have no idea what code you're checking against but it is surely not the upstream (that is, my) `develop` code. I'm guessing that you are testing your `develop`-branch which you modified in #31 where I told you that this is a bad idea...

I have no idea how your repository looks like at this moment, but make sure that your develop branch is the same as my develop branch and you base your work on top of that. [If you have added my repository as a remote](https://help.github.com/articles/configuring-a-remote-for-a-fork/) (let's say you've called it `upstream`), you can reset your develop branch with

```
git checkout develop
git reset --hard upstream/develop
```

After that, rebase your stuff on the true develop branch, not your modified one:

```
git checkout drag_and_drop
git rebase develop
```

**Comment by slowly-sloth** [comment]
I checkout your repo "https://github.com/denschub/firefox-tabgroups.git" to new project. Then 
- npm install
- jake lint

and i have something like:
http://imgur.com/ITaYahz

maybe is it issue with wrong eslint version?

**Comment by skyschub** [comment]
Which version of eslint do you run?

**Comment by slowly-sloth** [comment]
v1.10.3

**Comment by skyschub** [comment]
Why were you running 2.4.0 in https://github.com/denschub/firefox-tabgroups/pull/31#issuecomment-198244487? Upgrade your eslint.

**Comment by slowly-sloth** [comment]
Sorry for this mess :/ Now there are only warnings.

**Comment by skyschub** [comment]
In `develop`, there are only warnings about unused variables and two longer lines, you are right! IN your code however, a lot of stuff is wrong. Missing spaces, single quotes, ...

**Comment by slowly-sloth** [comment]
hmm... i've just checked my branches "drag_and_drop" & "develop" and there are only warnings.

**Comment by skyschub** [comment]
And because of there is a "warning" written in front of it, you decided to not fix the issues? I did not write rules for spacing and quotes for fun, you know?

```
data/components/group.js
  148:50  warning  Strings must use doublequote        quotes
  149:47  warning  Strings must use doublequote        quotes
  167:50  warning  Strings must use doublequote        quotes
  168:44  warning  Missing space before opening brace  space-before-blocks
  180:27  warning  Missing semicolon                   semi
  183:30  warning  Infix operators must be spaced      space-infix-ops
  183:35  warning  Missing space before opening brace  space-before-blocks

data/components/groupaddbutton.js
  13:22  warning  Block must not be padded by blank lines     padded-blocks
  58:27  warning  Missing semicolon                           semi
  61:30  warning  Infix operators must be spaced              space-infix-ops
  61:35  warning  Missing space before opening brace          space-before-blocks
  72:50  warning  Strings must use doublequote                quotes
  73:47  warning  Strings must use doublequote                quotes

data/components/grouplist.js
  13:24  warning  Extra space before value for key 'onTabDragStart'  key-spacing

data/components/tab.js
   9:19  warning  Unexpected space before function parentheses  space-before-function-paren
  39:27  warning  Unexpected space before function parentheses  space-before-function-paren
  49:26  warning  Unexpected space before function parentheses  space-before-function-paren
  53:32  warning  Strings must use doublequote                  quotes
  54:32  warning  Strings must use doublequote                  quotes
  62:31  warning  Unexpected space before function parentheses  space-before-function-paren
  66:32  warning  Strings must use doublequote                  quotes
  67:32  warning  Strings must use doublequote                  quotes
  73:4   warning  Unexpected trailing comma                     comma-dangle

index.js
  184:1  warning  Line 184 exceeds the maximum line length of 80  max-len
  203:1  warning  Line 203 exceeds the maximum line length of 80  max-len

lib/tabmanager.js
  101:1   warning  Line 101 exceeds the maximum line length of 80  max-len
  113:1   warning  Line 113 exceeds the maximum line length of 80  max-len
  212:48  warning  Missing space before opening brace              space-before-blocks
  215:28  warning  Infix operators must be spaced                  space-infix-ops
  215:31  warning  Missing space before opening brace              space-before-blocks
  216:60  warning  Infix operators must be spaced                  space-infix-ops
```

These are 31 style guide violations introduced by your changes. None of them is by any means discussable and I will not merge anything before those are fixed, which I've made clear before. A unique styling of code is an absolute basic skill and in this project, you do not even have to think about it, just read and fix what eslint shows you. If you are not interested in fixing these, I'll do it myself, but do not expect me to enjoy working with you if you do not care about the rules in open source projects.

**Comment by slowly-sloth** [comment]
my lint output for "drag_and_drop" branch:

```
D:\tk\dev\firefox\firefox-tabgroups\dist\data\action_creators.js
  1:7  warning  'ActionCreators' is defined but never used  no-unused-vars

D:\tk\dev\firefox\firefox-tabgroups\dist\data\components\app.js
  1:7  warning  'App' is defined but never used  no-unused-vars

D:\tk\dev\firefox\firefox-tabgroups\dist\data\components\group.js
  1:7  warning  'Group' is defined but never used  no-unused-vars

D:\tk\dev\firefox\firefox-tabgroups\dist\data\components\groupaddbutton.js
  1:7  warning  'GroupAddButton' is defined but never used  no-unused-vars

D:\tk\dev\firefox\firefox-tabgroups\dist\data\components\groupcontrols.js
  1:7  warning  'GroupControls' is defined but never used  no-unused-vars

D:\tk\dev\firefox\firefox-tabgroups\dist\data\components\grouplist.js
  1:7  warning  'GroupList' is defined but never used  no-unused-vars

D:\tk\dev\firefox\firefox-tabgroups\dist\data\components\tab.js
  1:7  warning  'Tab' is defined but never used  no-unused-vars

D:\tk\dev\firefox\firefox-tabgroups\dist\data\components\tablist.js
  1:7  warning  'TabList' is defined but never used  no-unused-vars

D:\tk\dev\firefox\firefox-tabgroups\dist\data\reducer.js
  5:7  warning  'Reducer' is defined but never used  no-unused-vars

D:\tk\dev\firefox\firefox-tabgroups\dist\index.js
  201:1  warning  Line 201 exceeds the maximum line length of 80  max-len

D:\tk\dev\firefox\firefox-tabgroups\dist\lib\tabmanager.js
  71:1  warning  Line 71 exceeds the maximum line length of 80  max-len
  83:1  warning  Line 83 exceeds the maximum line length of 80  max-len

? 12 problems (0 errors, 12 warnings)
```

just check for ex.

```
> data/components/tab.js
>  66:32  warning  Strings must use doublequote                  quotes
>  67:32  warning  Strings must use doublequote                  quotes
```

this is fixed (IMO)
https://github.com/xiio/firefox-tabgroups/blob/drag_and_drop/src/data/components/tab.js#L66-L67

**Comment by skyschub** [comment]
So, you've force pushed code in the meantime without telling me so. I'm sorry, I completely missed that. Sorry for the aggression, but I'm really tired of people ignoring rules. :cake: 

Your code looks really fine. I'll have to do some more careful testing to see if there are any edge cases left, but at this moment, I see none. I hope to get the testing done today!

A side node: After all those rebasings and stuff, the commit history got quite dirty. I'll squash your commits when merging so you don't have to deal with that, but GitHub has some great articles on Git and [the git book](https://git-scm.com/book/en/v2) is an amazing resource for learning your tools. Understanding git will make your life easier.

**Comment by slowly-sloth** [comment]
I'm sorry too for this mess. It's my first time with eslint. And i'm quite new in github collaboration. Thx for materials to learn. I promise next PR will be clear. Thx for your patience!

File: src/data/components/groupaddbutton.js [hunk 1]
```diff
@@ -21,5 +37,46 @@ const GroupAddButton = React.createClass({
...
+    var sourceGroup = event.dataTransfer.getData("tab/group");
+    var tabIndex = event.dataTransfer.getData("tab/index");
+
+    this.props.onDrop(
+      sourceGroup,
+      tabIndex
+    );
+  },
+
+  _dragDropCounter: 0
```
Comment (skyschub) [inline src/data/components/groupaddbutton.js]: This should be inside the state. Actually, the counter should be the only thing in the state since you do not really need `draggingOver`. In line 15, you could simply assign `this.state.dragDropCounter == 0`.

File: src/lib/tabmanager.js [hunk 2]
```diff
@@ -172,6 +234,7 @@ TabManager.prototype = {
 
     this._storage.removeGroupTabs(tabBrowser, groupID);
   }
+
 };
```
Comment (skyschub) [inline src/lib/tabmanager.js:224]: This newline is not needed. :)

File: src/lib/tabmanager.js [hunk 3]
```diff
@@ -66,6 +84,18 @@ TabManager.prototype = {
...
+
+  /**
+   * Show a given group.
+   *
+   * @param {ChromeWindow} chromeWindow
+   * @param {TabBrowser} tabBrowser
+   * @param {Number} groupID - the groupID
+   * @param {Number} tabIndex - the tab to activate
+   */
+  showGroup: function(chromeWindow, tabBrowser, groupID, tabIndex = 0) {
```
Comment (skyschub) [inline src/lib/tabmanager.js]: I do not understand your intentions. Why did you split `selectGroup` and `showGroup`? Besides the naming is now really confusing, `selectGroup` is simply calling `showGroup` but doing a check if it is actually necessary to switch. Why do you need to skip those checks? You're only calling `showGroup` inside `moveTabToGroup` and I see no reason to not call `selectGroup` there.

File: src/data/assets/css/groupspanel.css [hunk 4]
```diff
@@ -49,6 +50,16 @@ ul, li {
...
   }
 
+  .group.draggingOver {
+    border: 1px dashed #ccc;
+    background-color: rgba(0, 0, 0, 0.04);
+  }
+
+  .group.dragSourceGroup.draggingOver {
+    border: inherit;
+    background-color:inherit;
```
Comment (skyschub) [inline src/data/assets/css/groupspanel.css]: missing space after `:`.

File: src/data/components/group.js [hunk 5]
```diff
@@ -124,5 +137,54 @@ const Group = React.createClass({
...
-  }
+  },
+
+  handleGroupDrop: function(event) {
+    event.stopPropagation();
+
+    this.setState({draggingOver: false});
+    this._dragDropCounter = 0;
+
+    var sourceGroup = event.dataTransfer.getData("tab/group");
```
Comment (skyschub) [inline src/data/components/group.js]: `let`.

File: src/data/components/group.js [hunk 6]
```diff
@@ -124,5 +137,54 @@ const Group = React.createClass({
...
+    var tabIndex = event.dataTransfer.getData("tab/index");
```
Comment (skyschub) [inline src/data/components/group.js]: `let`.

File: src/data/components/group.js [hunk 7]
```diff
@@ -124,5 +137,54 @@ const Group = React.createClass({
...
+  handleGroupDragOver: function(event) {
+    event.stopPropagation();
+    event.preventDefault();
+  },
+
+  handleGroupDragEnter: function(event) {
+    event.stopPropagation();
+    event.preventDefault();
+
+    var sourceGroup = event.dataTransfer.getData("tab/group");
```
Comment (skyschub) [inline src/data/components/group.js]: `let`.

Comment (skyschub) [inline src/data/components/group.js]: Also, `sourceGroupId` would be a better name.

File: src/data/components/group.js [hunk 8]
```diff
@@ -124,5 +137,54 @@ const Group = React.createClass({
...
+    if (sourceGroup == this.props.group.id) {
+      this.setState({dragSourceGroup: true});
+    } else {
+      this.setState({dragSourceGroup: false});
+    }
```
Comment (skyschub) [inline src/data/components/group.js]:
``` js
this.setState({dragSourceGroup: sourceGroup == this.props.group.id});
```

File: src/lib/tabmanager.js (hunk 3 above)
Comment (slowly-sloth) [inline src/lib/tabmanager.js]: It was a concept at the beginning which i drop. I will move this back to `selectGroup`

**Comment by slowly-sloth** [comment]
I've fixed your review notes. i've found an issue with dragEnter dragLeave over elements with children. It should work now.

**Comment by skyschub** [comment]
I've merged this as 0d6300c845f648b27da015628cb2e63df32afb2b. Thank you very much for sticking with me! I'll probably release 0.4.0 somewhere later today so we can get this out as soon as possible. :)
